server/src/test/java/org/elasticsearch/index/query/SearchExecutionContextTests.java

```diff
@@ -388,18 +389,22 @@ public void testSyntheticSourceScriptLoading() throws IOException {
         MappingLookup lookup = MappingLookup.fromMapping(mapping);
 
         SearchExecutionContext sec = createSearchExecutionContext("index", "", lookup, Map.of());
+        assertTrue(sec.isSourceSynthetic());
+
+        MemoryIndex mi = new MemoryIndex();
+        mi.addField(new StringField("unloaded", "value", Field.Store.YES), null);
+        LeafReaderContext leafReaderContext = mi.createSearcher().getIndexReader().leaves().get(0);
 
-        // Attempting to access synthetic source via this context should throw an error
         SearchLookup searchLookup = sec.lookup();
-        Exception e = expectThrows(IllegalArgumentException.class, () -> searchLookup.getSource(null, 0));
-        assertThat(e.getMessage(), equalTo("Cannot access source from scripts in synthetic mode"));
+        assertNotNull(searchLookup.getSource(null, 0));
+        assertNotNull(searchLookup.getSource(leafReaderContext, 0));
 
         // Setting the source provider explicitly then gives us a new SearchLookup that can use source
-        Source source = Source.fromMap(Map.of("field", "value"), XContentType.JSON);
-        sec.setLookupProviders((ctx, doc) -> source, LeafFieldLookupProvider.fromStoredFields());
+        Source source1 = Source.fromMap(Map.of("field", "value"), XContentType.JSON);
```

Comment:
I think you can just remove this bit of the test; maybe add some testing that the Source returned earlier has the expected value?

Reply:
I removed this part, but it's tricky to get a unittest that evaluates a script - `TestRuntimeField` doesn't return actual values.. We get more coverage here from the yaml test, but I can try to subclass TestRuntimeField (or something along these lines) if you insist.

Reply:
I think you can just check that `Source.asMap().get("unloaded").get(0)` returns what you expect?  No need for scripting, just look at the constructed `Source` directly.

Reply:
I wish, `asMap` is null, I think we're getting into the "sourceLoader.requiredStoredFields().isEmpty()" path..

Reply:
That sounds like a bug?  It should be reconstructing the source for you, surely?

Reply:
Oh hang on, no I see what the issue is, you have an empty Mapping so there's nothing for it to base its reconstruction on.  I think you can add a field to the mapping, make sure that it's added as either or a stored field or a doc value field to the MemoryIndex, and then check the reconstruction.  In fact it's probably worth checking both types of reconstruction (stored field & doc value).

Reply:
I added a field to the mapping, but I still see an empty `storedFieldLoaders` in `SourceLoader.Synthetic`.. Updated the test, will keep looking.

Reply:
I get to see `asMap` containing a keyword but not the runtime field. I think the reason is that runtime fields are not added to the fields that the SyntheticFieldLoader tracks [1,2]. I can see in [2] that `extra` contains the SourceFieldMapper and `mappers` contains the KeywordFieldMapper; the runtime field is tracked under member `runtimeFields` but not added here..

I committed my latest changes, any further help here is appreciated.

[1] https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/mapper/Mapping.java#L128
[2]  https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java#L602

Reply:
Synthetic source works by rebuilding the source document using other index structures, but runtime fields don't have any index structures so they will indeed not work here.

I think this might be better tested if we can hook it into `MapperServiceTestCase`, specifically the `syntheticSource()` method on line 721.  If you can rewrite that method to use this new SourceProvider to load the Source, then it will get tested for every mapping type.

Reply:
Nice, updated.